cmdstanpy/utils.py

```diff
@@ -128,46 +128,127 @@ def check_csv(filename: str) -> Dict:
+        cur_pos = fp.tell()
         line = fp.readline().strip()
-        lineno += 1
-        while len(line) > 0 and line.startswith('#'):
-            line = fp.readline().strip()
+    fp.seek(cur_pos)
+    return lineno
+
+
+def scan_column_names(fp: TextIO, dict: Dict, lineno: int) -> int:
+    """
+    Parse column header into dict entries column_names
+    """
+    line = fp.readline().strip()
+    lineno += 1
+    names = line.split(',')
+    dict['column_names'] = tuple(names)
+    return lineno
+
+
+def scan_metric(fp: TextIO, dict: Dict, lineno: int) -> int:
+    """
+    Scan stepsize, metric from  stan_csv file comment lines,
+    set dict entries 'metric' and 'num_params'
+    """
+    if 'metric' not in dict:
+        dict['metric'] = 'diag_e'
+    metric = dict['metric']
+    line = fp.readline().strip()
+    lineno += 1
+    if not line == '# Adaptation terminated':
+        raise ValueError(
+            'line {}: expecting metric, '
+            'found:\n\t "{}"'.format(lineno, line))
+    line = fp.readline().strip()
+    lineno += 1
+    label, stepsize = line.split('=')
+    if not label.startswith('# Step size'):
+        raise ValueError(
+            'line {}: expecting stepsize, '
+            'found:\n\t "{}"'.format(lineno, line))
+    try:
+        float(stepsize.strip())
+    except ValueError:
+        raise ValueError(
+            'line {}: invalid stepsize: {}'.format(
+                lineno, stepsize))
+    line = fp.readline().strip()
+    lineno += 1
+    if not ((metric == 'diag_e' and
+                 line == '# Diagonal elements of inverse mass matrix:') or
+                 (metric == 'dense_e' and
+                      line == '# Elements of inverse mass matrix:')):
+        raise ValueError(
+            'line {}: invalid or missing mass matrix '
+            'specification'.format(lineno))
+    line = fp.readline().lstrip(' #\t')
+    lineno += 1
+    num_params = len(line.split(','))
+    dict['num_params'] = num_params
+    if metric == 'diag_e':
+        return lineno
+    else:
+        for i in range(1,num_params):
+            line = fp.readline().lstrip(' #\t')
             lineno += 1
-        while len(line) > 0 and not line.startswith('#'):
-            draws_found += 1
-            if len(line.split(',')) != num_cols:
+            if len(line.split(',')) != num_params:
                 raise ValueError(
-                    'file {}, at line {}: bad draw, expecting {} items, '
-                    'found {}'.format(
-                        filename, lineno, num_cols, len(line.split(',')))
-                )
-            line = fp.readline().strip()
-            lineno += 1
-    dict['draws'] = draws_found
-    return dict
+                    'line {}: invalid or missing mass matrix '
+                    'specification'.format(lineno))
+        return lineno
 
 
-def parse_header(line: str, dict: Dict) -> None:
+def scan_draws(fp: TextIO, dict: Dict, lineno: int) -> int:
```

Comment:
Do we really need to scan draws? (Maybe)

Reply:
I think that under certain circumstances when error is encountered in the `generated quantities` block, the sampler doesn't write the draw to the csv_file but does continue processing.  This isn't necessarily an error, nonetheless, the user needs to know if the sampler isn't behaving as advertised, especially if downstream analysis expects all chains to have the same shape.